fuzz/src/filter.rs

```diff
@@ -0,0 +1,105 @@
+use vortex::accessor::ArrayAccessor;
+use vortex::array::{BoolArray, PrimitiveArray, StructArray, VarBinArray};
+use vortex::validity::{ArrayValidity, Validity};
+use vortex::variants::StructArrayTrait;
+use vortex::{Array, ArrayDType, IntoArray, IntoArrayVariant};
+use vortex_dtype::{match_each_native_ptype, DType};
+
+pub fn filter_canonical_array(array: &Array, filter: &[bool]) -> Array {
+    match array.dtype() {
+        DType::Bool(_) => {
+            let bool_array = array.clone().into_bool().unwrap();
+            let vec_validity = bool_array
+                .logical_validity()
+                .into_array()
+                .into_bool()
+                .unwrap()
+                .boolean_buffer();
+            BoolArray::from_vec(
+                filter
+                    .iter()
+                    .zip(bool_array.boolean_buffer().iter())
+                    .filter(|(f, _)| **f)
+                    .map(|(_, v)| v)
+                    .collect::<Vec<_>>(),
+                Validity::from(
+                    filter
+                        .iter()
+                        .zip(vec_validity.iter())
+                        .filter(|(f, _)| **f)
+                        .map(|(_, v)| v)
+                        .collect::<Vec<_>>(),
+                ),
+            )
+            .into_array()
+        }
+        DType::Primitive(p, _) => match_each_native_ptype!(p, |$P| {
+            let primitive_array = array.clone().into_primitive().unwrap();
+            let vec_validity = primitive_array
+                .logical_validity()
+                .into_array()
+                .into_bool()
+                .unwrap()
+                .boolean_buffer();
+            PrimitiveArray::from_vec(
+                filter
+                    .iter()
+                    .zip(primitive_array.maybe_null_slice::<$P>().iter().copied())
+                    .filter(|(f, _)| **f)
+                    .map(|(_, v)| v)
+                    .collect::<Vec<_>>(),
+                Validity::from(
+                    filter
+                        .iter()
+                        .zip(vec_validity.iter())
+                        .filter(|(f, _)| **f)
+                        .map(|(_, v)| v)
+                        .collect::<Vec<_>>(),
+                ),
+            )
+            .into_array()
+        }),
+        DType::Utf8(_) | DType::Binary(_) => {
+            let utf8 = array.clone().into_varbin().unwrap();
+            let values = utf8
+                .with_iterator(|iter| {
+                    iter.zip(filter.iter())
+                        .filter(|(_, f)| **f)
+                        .map(|(v, _)| v.map(|u| u.to_vec()))
```

Comment:
nit - filter_map

Reply:
Clippy really wants you to use filter().map() instead of filter_map with bool::then